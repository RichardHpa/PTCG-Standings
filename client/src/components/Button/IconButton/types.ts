import type { ButtonProps } from '../types';
import type { Icon } from 'types/icons';

export interface IconButtonProps extends Omit<ButtonProps, 'children'> {
  alt: string;
  icon: Icon;
}
