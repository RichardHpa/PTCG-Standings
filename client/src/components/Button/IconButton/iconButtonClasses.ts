import { tw } from 'utils/tailwindClassName';

export const iconButtonClasses = {
  base: tw`inline-flex items-center justify-center rounded-lg p-2.5 text-center text-sm font-medium`,
  noPadding: tw`!p-0`,
  icon: {
    base: tw`inline-flex`,
    size: {
      xs: tw`h-4 w-4`,
      sm: tw`h-4 w-4`,
      base: tw`h-10 w-10`,
      lg: tw`h-6 w-6`,
      xl: tw`h-6 w-6`,
    },
  },
};
