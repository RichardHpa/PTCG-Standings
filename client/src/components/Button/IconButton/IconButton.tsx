import { cloneElement } from 'react';
import clsx from 'clsx';

import { iconButtonClasses } from './iconButtonClasses';
import { buttonClasses } from '../buttonClasses';

import type { FC } from 'react';
import type { IconButtonProps } from './types';

export const IconButton: FC<IconButtonProps> = ({
  color = 'primary',
  variant = 'contained',
  alt,
  icon,
  noPadding,
  size = 'base',
  disabled = false,
  ...props
}) => {
  const RenderedIcon = cloneElement(icon, {
    className: clsx(iconButtonClasses.icon.base),
    'aria-hidden': true,
    fill: 'currentColor',
  });

  return (
    <button
      type="button"
      {...props}
      className={clsx(
        iconButtonClasses.base,
        buttonClasses.variant[variant]['base'],
        buttonClasses.variant[variant][color].light,
        buttonClasses.variant[variant][color].dark,
        iconButtonClasses.icon.size[size],
        noPadding && iconButtonClasses.noPadding,
        disabled && buttonClasses.disabled,
      )}
    >
      {RenderedIcon}
      <span className="sr-only">{alt}</span>
    </button>
  );
};
