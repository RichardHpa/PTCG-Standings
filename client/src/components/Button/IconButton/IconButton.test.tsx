import { describe, expect, test, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { IconButton } from './IconButton';
import { SunIcon } from '@heroicons/react/16/solid';

describe('IconButton', () => {
  test('renders default IconButton', () => {
    render(<IconButton alt="test" icon={<SunIcon />} />);
    const button = screen.getByRole('button', { name: /test/i });
    expect(button).toBeInTheDocument();
  });

  test('renders IconButton with onClick', async () => {
    const onClick = vi.fn();
    render(<IconButton alt="test" icon={<SunIcon />} onClick={onClick} />);
    userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(onClick).toHaveBeenCalled());
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('renders IconButton with type', () => {
    render(<IconButton alt="test" icon={<SunIcon />} type="submit" />);
    expect(screen.getByRole('button')).toHaveAttribute('type', 'submit');
  });

  test.each([['contained'], ['outlined'], ['ghost']] as const)(
    'renders IconButton with variant %s',
    variant => {
      const { container } = render(
        <IconButton alt="test" icon={<SunIcon />} variant={variant} />,
      );
      expect(container).toMatchSnapshot();
    },
  );
});
