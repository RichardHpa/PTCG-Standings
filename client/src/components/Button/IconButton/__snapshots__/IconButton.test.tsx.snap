// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IconButton > renders IconButton with variant contained 1`] = `
<div>
  <button
    class="inline-flex h-10 w-10 items-center justify-center rounded-lg p-2.5 text-center text-sm font-medium bg-primary text-white hover:bg-primary-hover text-white dark:bg-primary-dark dark:hover:bg-primary-hoverDark"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="inline-flex"
      data-slot="icon"
      fill="currentColor"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 1a.75.75 0 0 1 .75.75v1.5a.75.75 0 0 1-1.5 0v-1.5A.75.75 0 0 1 8 1ZM10.5 8a2.5 2.5 0 1 1-5 0 2.5 2.5 0 0 1 5 0ZM12.95 4.11a.75.75 0 1 0-1.06-1.06l-1.062 1.06a.75.75 0 0 0 1.061 1.062l1.06-1.061ZM15 8a.75.75 0 0 1-.75.75h-1.5a.75.75 0 0 1 0-1.5h1.5A.75.75 0 0 1 15 8ZM11.89 12.95a.75.75 0 0 0 1.06-1.06l-1.06-1.062a.75.75 0 0 0-1.062 1.061l1.061 1.06ZM8 12a.75.75 0 0 1 .75.75v1.5a.75.75 0 0 1-1.5 0v-1.5A.75.75 0 0 1 8 12ZM5.172 11.89a.75.75 0 0 0-1.061-1.062L3.05 11.89a.75.75 0 1 0 1.06 1.06l1.06-1.06ZM4 8a.75.75 0 0 1-.75.75h-1.5a.75.75 0 0 1 0-1.5h1.5A.75.75 0 0 1 4 8ZM4.11 5.172A.75.75 0 0 0 5.173 4.11L4.11 3.05a.75.75 0 1 0-1.06 1.06l1.06 1.06Z"
      />
    </svg>
    <span
      class="sr-only"
    >
      test
    </span>
  </button>
</div>
`;

exports[`IconButton > renders IconButton with variant ghost 1`] = `
<div>
  <button
    class="inline-flex h-10 w-10 items-center justify-center rounded-lg p-2.5 text-center text-sm font-medium text-primary hover:bg-primary-hoverLight hover:text-primary dark:text-primary-dark dark:hover:bg-primary-hover dark:hover:bg-primary-hoverDark dark:hover:text-white"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="inline-flex"
      data-slot="icon"
      fill="currentColor"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 1a.75.75 0 0 1 .75.75v1.5a.75.75 0 0 1-1.5 0v-1.5A.75.75 0 0 1 8 1ZM10.5 8a2.5 2.5 0 1 1-5 0 2.5 2.5 0 0 1 5 0ZM12.95 4.11a.75.75 0 1 0-1.06-1.06l-1.062 1.06a.75.75 0 0 0 1.061 1.062l1.06-1.061ZM15 8a.75.75 0 0 1-.75.75h-1.5a.75.75 0 0 1 0-1.5h1.5A.75.75 0 0 1 15 8ZM11.89 12.95a.75.75 0 0 0 1.06-1.06l-1.06-1.062a.75.75 0 0 0-1.062 1.061l1.061 1.06ZM8 12a.75.75 0 0 1 .75.75v1.5a.75.75 0 0 1-1.5 0v-1.5A.75.75 0 0 1 8 12ZM5.172 11.89a.75.75 0 0 0-1.061-1.062L3.05 11.89a.75.75 0 1 0 1.06 1.06l1.06-1.06ZM4 8a.75.75 0 0 1-.75.75h-1.5a.75.75 0 0 1 0-1.5h1.5A.75.75 0 0 1 4 8ZM4.11 5.172A.75.75 0 0 0 5.173 4.11L4.11 3.05a.75.75 0 1 0-1.06 1.06l1.06 1.06Z"
      />
    </svg>
    <span
      class="sr-only"
    >
      test
    </span>
  </button>
</div>
`;

exports[`IconButton > renders IconButton with variant outlined 1`] = `
<div>
  <button
    class="inline-flex h-10 w-10 items-center justify-center rounded-lg p-2.5 text-center text-sm font-medium outline outline-1 hover:text-white text-primary outline-primary hover:bg-primary hover:outline-primary dark:text-primary-dark dark:outline-primary-dark hover:dark:bg-primary-dark"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="inline-flex"
      data-slot="icon"
      fill="currentColor"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 1a.75.75 0 0 1 .75.75v1.5a.75.75 0 0 1-1.5 0v-1.5A.75.75 0 0 1 8 1ZM10.5 8a2.5 2.5 0 1 1-5 0 2.5 2.5 0 0 1 5 0ZM12.95 4.11a.75.75 0 1 0-1.06-1.06l-1.062 1.06a.75.75 0 0 0 1.061 1.062l1.06-1.061ZM15 8a.75.75 0 0 1-.75.75h-1.5a.75.75 0 0 1 0-1.5h1.5A.75.75 0 0 1 15 8ZM11.89 12.95a.75.75 0 0 0 1.06-1.06l-1.06-1.062a.75.75 0 0 0-1.062 1.061l1.061 1.06ZM8 12a.75.75 0 0 1 .75.75v1.5a.75.75 0 0 1-1.5 0v-1.5A.75.75 0 0 1 8 12ZM5.172 11.89a.75.75 0 0 0-1.061-1.062L3.05 11.89a.75.75 0 1 0 1.06 1.06l1.06-1.06ZM4 8a.75.75 0 0 1-.75.75h-1.5a.75.75 0 0 1 0-1.5h1.5A.75.75 0 0 1 4 8ZM4.11 5.172A.75.75 0 0 0 5.173 4.11L4.11 3.05a.75.75 0 1 0-1.06 1.06l1.06 1.06Z"
      />
    </svg>
    <span
      class="sr-only"
    >
      test
    </span>
  </button>
</div>
`;
